### DISCUSSION TITLE
Question about SRP service creation

### DISCUSSION THREAD

**Discussion by OlivierGre** [openthread/openthread#8320] [Q&A]
I have recently changed my application to use the persistent memory. By this way, if I commission a Thread device and reboot it, it will be able to reconnect to the Thread network. I have implemented it and it works.

This leads me to some questions about SRP:
Once the device reconnected to Thread network, I want to ensure that it indicates its presence via SRP. 

2 cases are possible: 
- The SRP service might still be present (coming from a recent execution of the device) 
- or it might not. 

I'm wondering what is the nice way to handle this. 
I have thought to the following possibilities:

Solution 1: 
Once the device reconnected on Thread network, use  otSrpClientGetServices() to know if a SRP service with a special name is present (the name if the name of my service followed by the EUI64 of the device). 

Solution 2:
Try to do the SRP registration every times. If the SRP is already present, it will fail with the error OT_ERROR_DUPLICATED.
In that case I will probably have to cancel the registration otherwise there is an automatic retry.

Solution 3:
Once the device reconnected on Thread network, use  otSrpClientGetServices() and otSrpClientRemoveService() to remove a possibly existing service with the same name and then recreate it.

Can you please tell me what is the recommended way?

Is it a problem if all my devices use the same SRP Host name but a different SRP Service name? As indicated above the service name would contain the EUI64 of the device.
Exemple:
Host name: "ot-st25-host"
Service Instance name: "ot-st25-service-0080e105005ec74a"

Thank you

**Comment by OlivierGre** [comment]
I have tried to use  otSrpClientGetServices()  to know the currently existing SRP services but it doesn't work: It is returning NULL (whereas I can see with Service Browser that the SRP service is present). I suspect that otSrpClientGetServices() can only be used to know the service started during current execution. I have rebooted the device so that's probably why it is not aware any more about this service.

It that's correct, I think that only Solution 2 remains...

**Comment by OlivierGre** [comment]
I have also tried to call otSrpClientRemoveHostAndServices() to remove an existing SRP service but it doesn't work. It fails with this error:

`[00:00:01.998] otSrpClientRemoveHostAndServices failed with error 24`

**Comment by jwhui** [comment]
An SRP client can unregister all host and services from the SRP server using [otSrpClientRemoveHostAndServices()](https://openthread.io/reference/group/api-srp#otsrpclientremovehostandservices).

Note, however, that SRP cryptographically signs all messages with a key. OpenThread's SRP Client stores the key in non-volatile memory, so this should work after a simple reboot. However, if you factory reset the device or otherwise delete the SRP client key from non-volatile storage in any way, the SRP client will not be able to update information that it had previously registered with the SRP server.

**Reply by OlivierGre** [reply]
Ok understood, thank you. I have not erased the NVM intentionally but it may have happened automatically if I have launched a new commissioning while the device was still connected. I will take care next time.

Isn't the function [otSrpClientRemoveHostAndServices()](https://openthread.io/reference/group/api-srp#otsrpclientremovehostandservices) a bit strong? I mean, it will clear all hosts and services so it will impact other devices in the network. Is there a way to clear only the host and the service related to the current device?

**Reply by jwhui** [reply]
An SRP Client can only manage information that it registered. Remember that the SRP Client cryptographically signs its updates. So [otSrpClientRemoveHostAndServices()](https://openthread.io/reference/group/api-srp#otsrpclientremovehostandservices) only impacts information registered by the device itself.

**Reply by OlivierGre** [reply]
Oh, ok, perfect. Thanks to both of you for these explanations.

**Comment by abtink** [comment]
About the names, 
- You want to use unique/different host name per device (per SRP client).
- Service name can be the same, but then service instance name should be unique per device.

The first device (SRP client) which registers a host name will be considered as the owner of that name by sever and only the same device (using the same SRP key) can change the registration  (add/remove services, IPv6 addresses, or remove the host) for the same host name. If later another device tries to register the same host name it will be rejected by server.

**Reply by OlivierGre** [reply]
Good, thank you for this explanation. So I will use:
Host name: "ot-st25-host-0080e105005ec74a"
Service Instance name: "ot-st25-service-0080e105005ec74a"

**Comment by OlivierGre** [comment]
> An SRP client can unregister all host and services from the SRP server using [otSrpClientRemoveHostAndServices()](https://openthread.io/reference/group/api-srp#otsrpclientremovehostandservices).
> 
> Note, however, that SRP cryptographically signs all messages with a key. OpenThread's SRP Client stores the key in non-volatile memory, so this should work after a simple reboot. However, if you factory reset the device or otherwise delete the SRP client key from non-volatile storage in any way, the SRP client will not be able to update information that it had previously registered with the SRP server.

@jwhui I don't know what I could have done wrong but what you described doesn't work for me.
I'm doing the things in the following order:
1) Perform the commissioning of a Thread device. 
2) Save the OT data in NVM. 
3) Start SRP.

I then click on reset button. 
The device reconnects to thread network and then calls the function to start SRP. This function first call otsrpclientremovehostandservices() but it fails with the error 24:
[00:00:21.371] otSrpClientRemoveHostAndServices failed with error 24

Do you know what I could have done wrong? Is it necessary to save the OT data in NVM after the creation of the SRP service? 
If yes, the sequence would be:
1) Perform the commissioning of a Thread device. 
2) Save the OT data in NVM (this will allow to reconnect to the Thread network).
3) Start SRP.
4) Save the OT data in NVM (this will allow to get the cryptographic keys if we want to remove the SRP service later).

That's just a guess...

**Reply by abtink** [reply]
The error 24 is `OT_ERROR_ALREADY`. 

It means there was nothing to remove.
- You can check the documentation of API and its returned errors in https://openthread.io/reference/group/api-srp#otsrpclientremovehostandservices.
- Did you set the host name that you want to remove after restart? I guess maybe not.

**Reply by OlivierGre** [reply]
Oh, sorry for that!!! I was wrong indeed.

> * Did you set the host name that you want to remove after restart? I guess maybe not.

In the function in charge of starting the SRP service, I do:

```
    // Stop SRP services that may have been started previously by this device
    StopSRPService();    // This function will call otSrpClientRemoveHostAndServices()

    SrpSetHostName();
    SrpSetHostAddress();
    SrpServiceAdd();
    SrpEnableAutostart();
```
According to your message, it seems that I should set the host name before calling StopSRPService().

**Reply by abtink** [reply]
Yes. Correct. You need to specify which host name you want to remove. I would suggest, doing `SrpSetHostName();` and maybe even 
 `SrpSetHostAddress();` before the function that calls `otSrpClinetRemoveHostAndServices()`.
Also you may want to invoke it with `aSendUnregToServer` as `true` and also `aRemoveKeyLease` as `true`.

**Reply by OlivierGre** [reply]
Ok, understood, thank you. For aSendUnregToServer and aRemoveKeyLease, this is what I'm doing.
